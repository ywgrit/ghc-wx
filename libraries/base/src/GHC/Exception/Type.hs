{-# LANGUAGE Safe #-}
{-# OPTIONS_HADDOCK not-home #-}

-- |
-- Module      :  GHC.Exception.Type
-- Copyright   :  (c) The University of Glasgow, 1998-2002
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  cvs-ghc@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Exceptions and exception-handling functions.
--

module GHC.Exception.Type
       ( Exception(..)    -- Class
       , SomeException(..), ArithException(..)
       , divZeroException, overflowException, ratioZeroDenomException
       , underflowException
       ) where

import GHC.Internal.Exception.Type
