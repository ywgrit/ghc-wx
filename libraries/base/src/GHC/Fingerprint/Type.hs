{-# LANGUAGE Safe #-}

-- |
--
-- Module      :  GHC.Fingerprint.Type
-- Copyright   :  (c) The University of Glasgow, 1994-2023
-- License     :  see libraries/base/LICENSE
--
-- Maintainer  :  ghc-devs@haskell.org
-- Stability   :  internal
-- Portability :  non-portable (GHC extensions)
--
-- Fingerprints for recompilation checking and ABI versioning, and
-- implementing fast comparison of Typeable.
--
-- /The API of this module is unstable and not meant to be consumed by the general public./
-- If you absolutely must depend on it, make sure to use a tight upper
-- bound, e.g., @base < 4.X@ rather than @base < 5@, because the interface can
-- change rapidly without much warning.

module GHC.Fingerprint.Type
    (Fingerprint(..)
     ) where

import GHC.Internal.Fingerprint.Type