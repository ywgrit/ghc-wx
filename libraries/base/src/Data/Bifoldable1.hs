-- |
-- Copyright: Edward Kmett, Oleg Grenrus
-- License: BSD-3-Clause
--

{-# LANGUAGE NoImplicitPrelude #-}
{-# LANGUAGE Safe              #-}

module Data.Bifoldable1 where

import Control.Applicative (Const (..))
import Data.Bifoldable     (Bifoldable (..))
import Data.Semigroup      (Arg (..), Semigroup (..))
import Prelude             (Either (..), id)

class Bifoldable t => Bifoldable1 t where
     bifold1 :: Semigroup m => t m m -> m
     bifold1 = bifoldMap1 id id
     {-# INLINE bifold1 #-}

     bifoldMap1 :: Semigroup m => (a -> m) -> (b -> m) -> t a b -> m

instance Bifoldable1 Arg where
    bifoldMap1 f g (Arg a b) = f a <> g b

instance Bifoldable1 Either where
    bifoldMap1 f _ (Left a) = f a
    bifoldMap1 _ g (Right b) = g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 (,) where
    bifoldMap1 f g (a, b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 ((,,) x) where
    bifoldMap1 f g (_,a,b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 ((,,,) x y) where
    bifoldMap1 f g (_,_,a,b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 ((,,,,) x y z) where
    bifoldMap1 f g (_,_,_,a,b) = f a <> g b
    {-# INLINE bifoldMap1 #-}

instance Bifoldable1 Const where
    bifoldMap1 f _ (Const a) = f a
    {-# INLINE bifoldMap1 #-}
