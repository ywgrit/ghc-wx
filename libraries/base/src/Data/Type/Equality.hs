{-# LANGUAGE Safe #-}

{-# LANGUAGE ExplicitNamespaces #-}

-- |
--
-- Module      :  Data.Type.Equality
-- License     :  BSD-style (see the LICENSE file in the distribution)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  not portable
--
-- Definition of propositional equality @(':~:')@. Pattern-matching on a variable
-- of type @(a ':~:' b)@ produces a proof that @a '~' b@.
--
-- @since 4.7.0.0

module Data.Type.Equality
    (-- *  The equality types
     type (~),
     type (~~),
     (:~:)(..),
     (:~~:)(..),
     -- *  Working with equality
     sym,
     trans,
     castWith,
     gcastWith,
     apply,
     inner,
     outer,
     -- *  Inferring equality from other types
     TestEquality(..),
     -- *  Boolean type-level equality
     type (==)
     ) where

import GHC.Internal.Data.Type.Equality
