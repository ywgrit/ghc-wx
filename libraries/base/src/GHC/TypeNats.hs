{-# LANGUAGE Safe #-}
{-# LANGUAGE ExplicitNamespaces #-}
{-# LANGUAGE PatternSynonyms #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE NoStarIsType #-}

-- |
-- This module is an internal GHC module.  It declares the constants used
-- in the implementation of type-level natural numbers.  The programmer interface
-- for working with type-level naturals should be defined in a separate module.
--
-- @since 4.10.0.0
--

module GHC.TypeNats
    (-- *  Nat Kind
     Natural,
     Nat,
     -- *  Linking type and value level
     KnownNat(natSing),
     natVal,
     natVal',
     SomeNat(..),
     someNatVal,
     sameNat,
     decideNat,
     -- **  Singleton values
     SNat,
     pattern SNat,
     fromSNat,
     withSomeSNat,
     withKnownNat,
     -- *  Functions on type literals
     type (<=),
     type (<=?),
     type (+),
     type (*),
     type (^),
     type (-),
     CmpNat,
     cmpNat,
     Div,
     Mod,
     Log2
     ) where

import GHC.Internal.TypeNats
