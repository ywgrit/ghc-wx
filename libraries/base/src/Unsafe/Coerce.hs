{-# LANGUAGE MagicHash #-}

module Unsafe.Coerce
    (unsafeCoerce,
     unsafeCoerceUnlifted,
     unsafeCoerceAddr,
     unsafeEqualityProof,
     UnsafeEquality(..),
     unsafeCoerce#
     ) where

import GHC.Internal.Unsafe.Coerce
