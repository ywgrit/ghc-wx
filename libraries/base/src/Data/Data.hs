{-# LANGUAGE Safe #-}

-- |
-- Module      :  Data.Data
-- Copyright   :  (c) The University of Glasgow, CWI 2001--2004
-- License     :  BSD-style (see the file libraries/base/LICENSE)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  non-portable (local universal quantification)
--
-- This module provides the 'Data' class with its primitives for
-- generic programming, along with instances for many datatypes. It
-- corresponds to a merge between the previous "Data.Generics.Basics"
-- and almost all of "Data.Generics.Instances". The instances that are
-- not present in this module were moved to the
-- @Data.Generics.Instances@ module in the @syb@ package.
--
-- \"Scrap your boilerplate\" --- Generic programming in Haskell.  See
-- <https://wiki.haskell.org/Research_papers/Generics#Scrap_your_boilerplate.21>.
--

module Data.Data (

        -- * Module Data.Typeable re-exported for convenience
        module Data.Typeable,

        -- * The Data class for processing constructor applications
        Data(
                gfoldl,
                gunfold,
                toConstr,
                dataTypeOf,
                dataCast1,      -- mediate types and unary type constructors
                dataCast2,      -- mediate types and binary type constructors
                -- Generic maps defined in terms of gfoldl
                gmapT,
                gmapQ,
                gmapQl,
                gmapQr,
                gmapQi,
                gmapM,
                gmapMp,
                gmapMo
            ),

        -- * Datatype representations
        DataType,       -- abstract
        -- ** Constructors
        mkDataType,
        mkIntType,
        mkFloatType,
        mkCharType,
        mkNoRepType,
        -- ** Observers
        dataTypeName,
        DataRep(..),
        dataTypeRep,
        -- ** Convenience functions
        repConstr,
        isAlgType,
        dataTypeConstrs,
        indexConstr,
        maxConstrIndex,
        isNorepType,

        -- * Data constructor representations
        Constr,         -- abstract
        ConIndex,       -- alias for Int, start at 1
        Fixity(..),
        -- ** Constructors
        mkConstr,
        mkConstrTag,
        mkIntegralConstr,
        mkRealConstr,
        mkCharConstr,
        -- ** Observers
        constrType,
        ConstrRep(..),
        constrRep,
        constrFields,
        constrFixity,
        -- ** Convenience function: algebraic data types
        constrIndex,
        -- ** From strings to constructors and vice versa: all data types
        showConstr,
        readConstr,

        -- * Convenience functions: take type constructors apart
        tyconUQname,
        tyconModule,

        -- * Generic operations defined in terms of 'gunfold'
        fromConstr,
        fromConstrB,
        fromConstrM

  ) where

import GHC.Internal.Data.Data
import Data.Typeable
