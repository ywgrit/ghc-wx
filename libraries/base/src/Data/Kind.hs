{-# LANGUAGE Trustworthy #-}

-- |
--
-- Module      :  Data.Kind
-- License     :  BSD-style (see the LICENSE file in the distribution)
--
-- Maintainer  :  libraries@haskell.org
-- Stability   :  stable
-- Portability :  not portable
--
-- Basic kinds
--
-- @since 4.9.0.0

module Data.Kind
    (Type,
     Constraint,
     FUN
     ) where

import GHC.Num.BigNat () -- for build ordering (#23942)
import GHC.Prim (FUN)
import GHC.Types (Type, Constraint)
