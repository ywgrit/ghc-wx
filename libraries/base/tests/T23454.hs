{-# LANGUAGE DataKinds #-}
{-# LANGUAGE TypeFamilies #-}

module T23454 where

import Data.Coerce (coerce)
import Data.Kind (Type)
import Data.Type.Equality
import GHC.TypeNats

bogus :: forall a b . KnownNat a => a :~: b
bogus = case testEquality (SNat @a) (coerce (SNat @a) :: SNat b) of
          Just r  -> r
          Nothing -> error "bug fixed"

type G :: Nat -> Type -> Type -> Type
type family G n s t where
  G 0 s _ = s
  G _ _ t = t

newtype N n s t = MkN { unN :: G n s t }

oops :: forall b s t . N 0 s t -> N b s t
oops x = gcastWith (bogus @0 @b) x

unsafeCoerce :: s -> t
unsafeCoerce x = unN (oops @1 (MkN x))
