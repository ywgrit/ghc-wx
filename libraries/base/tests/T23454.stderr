
T23454.hs:12:38: error: [GHC-25897]
    • Couldn't match type ‘a’ with ‘b’ arising from a use of ‘coerce’
      ‘a’ is a rigid type variable bound by
        the type signature for:
          bogus :: forall (a :: Nat) (b :: Nat). KnownNat a => a :~: b
        at T23454.hs:11:1-43
      ‘b’ is a rigid type variable bound by
        the type signature for:
          bogus :: forall (a :: Nat) (b :: Nat). KnownNat a => a :~: b
        at T23454.hs:11:1-43
    • In the second argument of ‘testEquality’, namely
        ‘(coerce (SNat @a) :: SNat b)’
      In the expression:
        testEquality (SNat @a) (coerce (SNat @a) :: SNat b)
      In the expression:
        case testEquality (SNat @a) (coerce (SNat @a) :: SNat b) of
          Just r -> r
          Nothing -> error "bug fixed"
    • Relevant bindings include
        bogus :: a :~: b (bound at T23454.hs:12:1)
